File: sections/14_core_patterns_mixed_contest/problems/t_minimum_refueling/solution.py
import heapq
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    destination, reachable, n = data[:3]
    stations = sorted(zip(data[3::2], data[4::2]))
    available = []
    index = stops = 0
    while reachable < destination:
        while index < n and stations[index][0] <= reachable:
            heapq.heappush(available, -stations[index][1])
            index += 1
        if not available:
            print(-1)
            return
        reachable -= heapq.heappop(available)
        stops += 1
    print(stops)


if __name__ == "__main__":
    main()

File: sections/14_core_patterns_mixed_contest/problems/t_minimum_refueling/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    long long destination, reachable;
    int n;
    cin >> destination >> reachable >> n;
    vector<pair<long long, long long>> stations(n);
    for (auto &[position, fuel] : stations) cin >> position >> fuel;
    sort(stations.begin(), stations.end());
    priority_queue<long long> available;
    int index = 0, stops = 0;
    while (reachable < destination) {
        while (index < n && stations[index].first <= reachable) {
            available.push(stations[index].second);
            ++index;
        }
        if (available.empty()) {
            cout << -1 << '\n';
            return 0;
        }
        reachable += available.top();
        available.pop();
        ++stops;
    }
    cout << stops << '\n';
}
